### PR TITLE
[CONTINT-4419] update batching logic to not split on 429

### PR DESCRIPTION
**Pull request by ewoodthomas**
### What does this PR do?

Prevents batches that just encountered a 429 rate limit error from splitting up the batch. Also migrated error message constants to distinct error types to allow for easier error recognition.

### Motivation

https://datadoghq.atlassian.net/browse/CONTINT-4419

Rate Limit errors occur fairly regularly. Current batching logic will split up the batch no matter what the error is. However for 429 rate limit errors, this only makes the problem worse by increasing calls to the API.

If any metric in the batch has a rate limit error, this error is applied to all metrics in the batch. Metrics with a rate limit error are not split off from the batch.

### Describe how to test/QA your changes

I did not want to invoke a rate limit error on the ddev API. 

However, I temporarily changed the logic for rate limit error detection from `strings.Contains(err.Error(), "429 Too Many Requests")` to `strings.Contains(err.Error(), "400 Bad Request")` to check how the batches are handled.

I deployed the agent, HPA, and a DatadogMetric configured as follows:

```
apiVersion: datadoghq.com/v1alpha1
kind: DatadogMetric
metadata:
  name: nginx
spec:
  query: ""
...
```
The empty query throws a 400 Bad Request error. Through some logs that I temporarily added, I was able to verify that the batch is not split.

### Additional Notes

May also want to add 422 Unprocessable Entity errors to this logic in the future.

The unit tests set the actual and expected errors to be equal to each other before asserting that they are equal. Fixed this so error messages are tested in the unit tests.

### CODE REVIEW

File: pkg/clusteragent/autoscaling/externalmetrics/metrics_retriever.go [hunk 1]
```diff
@@ -19,11 +19,11 @@ import (
 
 const (
 	invalidMetricErrorMessage                  string = "%v, query was: %s"
-	invalidMetricErrorWithRetriesMessage       string = "%v, query was: %s, will retry after %s."
+	invalidMetricErrorWithRetriesMessage       string = "%v, query was: %s, will retry after %s"
 	invalidMetricOutdatedErrorMessage          string = "Query returned outdated result, check MaxAge setting, query: %s"
-	invalidMetricNotFoundErrorMessage          string = "Unexpected error, query data not found in result, query: %s"
-	invalidMetricGlobalErrorMessage            string = "Global error (all queries) from backend, invalid syntax in query? Check Cluster Agent leader logs for details"
-	invalidMetricGlobalErrorWithRetriesMessage string = "Global error (all queries, batch size %d) from backend, invalid syntax in query? Check Cluster Agent leader logs for details. Will retry after %s."
+	invalidMetricNotFoundErrorMessage          string = "Unexpected error, query data not found in result, query: %s. Query error: %v"
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/externalmetrics/metrics_retriever.go:24]: I'm not sure sure if we want to leak internal errors from API in the error message. You may simply add dedicated error message based on error returned by the backend.

File: pkg/clusteragent/autoscaling/externalmetrics/metrics_retriever.go [hunk 2]
```diff
@@ -75,7 +75,8 @@ func (mr *MetricsRetriever) Run(stopCh <-chan struct{}) {
 }
 
 func (mr *MetricsRetriever) retrieveMetricsValues() {
-	if mr.splitBatchBackoffOnErrors {
+	// If any metric encountered a Rate Limit error, do not split up the batch
+	if mr.splitBatchBackoffOnErrors && !mr.store.ContainsErrorType(autoscalers.IsRateLimitError) {
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/externalmetrics/metrics_retriever.go:79]: With `ContainsErrorType` call you're already going through all the store at once, while you could check in the second `GetFiltered` if `datadogMetric.Error` is a Rate Limit error or not

File: pkg/clusteragent/autoscaling/externalmetrics/metrics_retriever.go [hunk 3]
```diff
@@ -160,33 +167,38 @@ func (mr *MetricsRetriever) retrieveMetricsValuesSlice(datadogMetrics []model.Da
...
 			} else {
 				// This should never happen as `QueryExternalMetric` is filling all missing series
 				// if no global error.
-				datadogMetricFromStore.Error = log.Errorf(invalidMetricNotFoundErrorMessage, query)
+				datadogMetricFromStore.Error = NewInvalidMetricNotFoundError(query)
+				log.Error(datadogMetricFromStore.Error)
 			}
 		}
+		// If there is a rate limit error, apply the rate limit error to all datadogMetrics
+		if rateLimitErr != nil {
```
Comment (vboulineau) [inline pkg/clusteragent/autoscaling/externalmetrics/metrics_retriever.go:198]: Not sure this is the right location. You'll still go through previous `else {}` that invalidates the DDM (which it should not). I believe you need some dedicated code either in an `else if {}` or in the large `else {}`

**Review by kangyili — APPROVED** [review]

**Review by vboulineau — APPROVED** [review]

**Review by gabedos — APPROVED** [review]
The change from using const error messages to creating error structs is great and it lends well to batching rateLimitErrors into the valid batch. Looks good to me.

Would just check the lint error:
pkg/clusteragent/autoscaling/externalmetrics/errors.go:120:9: S1039: unnecessary use of fmt.Sprintf
	return fmt.Sprintf("Global error: Datadog API rate limit exceeded")

File: pkg/clusteragent/autoscaling/externalmetrics/errors.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	}
+}
+
+// RateLimitError represents a rate limit error.
+type RateLimitError struct {
+	Err error
+}
+
+func (e *RateLimitError) Error() string {
+	return fmt.Sprintf("Global error: Datadog API rate limit exceeded")
```
Comment (jennchenn) [inline pkg/clusteragent/autoscaling/externalmetrics/errors.go:120]:
```suggestion
	return "Global error: Datadog API rate limit exceeded"
```

**Pull request merged**